### DISCUSSION TITLE
Bug: Prettier runs fine, but eslint reports problems from `prettier/prettier` when used in `overwrites` section of eslint config

### DISCUSSION THREAD

**Discussion by gossi** [eslint/eslint#17158] [Q&A] [answered]
### Environment

Node version: `v18.16.0`
npm version: `9.5.1`
Local ESLint version: `8.23.1`
Global ESLint version: -
Operating System: `macOS 12.4`

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

This is from my `frontend-config` repo (or the branch with the next version of it):

- [prettier config](https://github.com/gossi/frontend-configs/blob/2f9715120e/packages/prettier/index.js) 
- [eslint config](https://github.com/gossi/frontend-configs/tree/2f9715120e5c60f11f12518d41a06a3d09f399f9/packages/eslint) - this is quite complex, I gonna provide the relevant `--print-config` at the end of this post

As this is fairly complex, I have extracted the minimal example at the end of the post of this issue.
</details>

[`greeting.gts`](https://github.com/gossi/frontend-configs/blob/2f9715120e/testing/ember-addon-v2-gts/src/components/greeting.gts)

```gts
import Component from '@glimmer/component';

export interface GreetingSignature {
  Element: HTMLSpanElement;
  Args: {
    hello: string;
    to: string;
  };
}

export default class Greeting extends Component<GreetingSignature> {
  get to() {
    return `${this.args.to[0]?.toUpperCase()}${this.args.to.slice(1).toLowerCase()}`;
  }

  <template>
    <span ...attributes>{{@hello}} to {{this.to}}</span>
  </template>
}
```

### What did you expect to happen?

`eslint` and `prettier` to be happy with the formatting with the given config.

### What actually happened?

prettier is fine:

```sh
❯ pnpm prettier -c src/components/greeting.gts
Checking formatting...
All matched files use Prettier code style!
```

eslint complains that prettier is wrong:

```sh
❯ pnpm eslint src/components/greeting.gts

/Users/thomas/code/gossi/frontend-configs/testing/ember-addon-v2-gts/src/components/greeting.gts
  16:24  error  Replace `⏎····<span·...attributes>{{@hello}}·to·{{this.to}}</span>⏎··` with `<span·...attributes>{{@hello}}·to·{{this.to}}</span>`  prettier/prettier

✖ 1 problem (1 error, 0 warnings)
```

I could figure, this is due to the issue that the prettier config lives within an `overwrites` field (read on).

### Link to Minimal Reproducible Example

https://github.com/gossi/frontend-configs/tree/2f9715120e5c60f11f12518d41a06a3d09f399f9

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

In order to access the repro example above (Assuming you have `pnpm` installed):

```sh
git clone ...
pnpm install
cd testing/ember-addon-v2-gts
pnpm exec prettier
pnpm lint
```

spot the difference. Below are configs to put into the `testing/ember-addon-v2-gts/.eslintrc.js` to make the problem visible.

---

More notes on this:

- `gts` is the extension of ember's single file component (much like `.vue` and `.svelte`)
- There is a prettier plugin which contains the parser for this file format: [`prettier-plugin-ember-template-tag`](https://github.com/gitKrystan/prettier-plugin-ember-template-tag)
- [`prettier/prettier`](https://github.com/gossi/frontend-configs/blob/2f9715120e/packages/eslint/configs/rules/prettier.js) config reads in `.prettierrc.js` (see below)

**What debugging efforts, have I tried?**

I applied some other configuration configurations to trigger different results, which I did so here is how it went:

1. I have used `--print-config` to print the config, copied it over to `.eslintrc.js` and removed parts until eslint was fine validating. So, it apparently is a config problem (and not connected to the plugins mentioned before)

2. With that working config, I did one more `--print-config` to have a short reference config, that worked. (see first file below)

3. I then did `console.log(JSON.stringify(configs.ember()))` to get a compiled lengthy config. I reduced it step-by-step until it matched my reference config, but it wasn't validating :/ (see second file below)

4. I then realized this: The config that worked, had everything defined in the root object, the one that didn't was using `overwrites`. So, apparently `prettier/prettier` works different, when it is configured in the root, rather than in the overwrite. (see second file below)

---

In essence, the short config that works (output from `--print-config`):

```js
module.exports = {
  env: {
    browser: true,
    es2020: true,
    es6: true
  },
  globals: {
    __GLIMMER_TEMPLATE: 'readonly'
  },
  parser: '@typescript-eslint/parser',
  // extends: ['plugin:@typescript-eslint/recommended'],

  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2020,
    ecmaFeatures: {
      legacyDecorators: true
    },
    babelOptions: {
      configFile:
        '/Users/thomas/code/gossi/frontend-configs/node_modules/.pnpm/eslint-plugin-decorator-position@5.0.2_@babel+eslint-parser@7.21.3_eslint@8.38.0/node_modules/eslint-plugin-decorator-position/babel.config.cjs'
    }
  },
  plugins: ['@typescript-eslint', 'ember', 'prettier', 'decorator-position', 'simple-import-sort'],
  extends: ['plugin:@typescript-eslint/recommended'],
  rules: {
    'prettier/prettier': [
      'error',
      {
        singleQuote: true,
        trailingComma: 'none',
        printWidth: 100,
        plugins: ['prettier-plugin-ember-template-tag'],
        overrides: [
          {
            files: ['*.md', '*.hbs'],
            options: {
              singleQuote: false
            }
          }
        ]
      }
    ]
  }
};
```

and this is where the problem becomes visible:

```diff
module.exports = {
  env: {
    browser: true,
    es2020: true,
    es6: true
  },
  globals: {
    __GLIMMER_TEMPLATE: 'readonly'
  },
  // root: true,
+  overrides: [
+    {
      parser: '@typescript-eslint/parser',
      // extends: ['plugin:@typescript-eslint/recommended'],

      parserOptions: {
        sourceType: 'module',
        ecmaVersion: 2020,
        ecmaFeatures: {
          legacyDecorators: true
        },
        babelOptions: {
          configFile:
            '/Users/thomas/code/gossi/frontend-configs/node_modules/.pnpm/eslint-plugin-decorator-position@5.0.2_@babel+eslint-parser@7.21.3_eslint@8.38.0/node_modules/eslint-plugin-decorator-position/babel.config.cjs'
        }
      },
      plugins: [
        '@typescript-eslint',
        'ember',
        'prettier',
        'decorator-position',
        'simple-import-sort'
      ],
      extends: ['plugin:@typescript-eslint/recommended'],
      rules: {
        'prettier/prettier': [
          'error',
          {
            singleQuote: true,
            trailingComma: 'none',
            printWidth: 100,
            plugins: ['prettier-plugin-ember-template-tag'],
            overrides: [
              {
                files: ['*.md', '*.hbs'],
                options: {
                  singleQuote: false
                }
              }
            ]
          }
        ]
+      },
+      files: ['{src,app,addon,addon-test-support,tests,types}/**/*.gts']
+    }
+  ]
};
```

=> My senses tell me, when the config is put into `overwrites`, something is handled differently in unexpected ways

---

This was originally reported over at prettier/eslint-plugin-prettier#557.

**Comment by Rec0iL99** [comment]
Hi @gossi, thanks for the issue. I was able to reproduce the issue.

**Case 1:**

I removed the original contents of the `.eslintrc.js` file and replaced it with the below config content:

```js
module.exports = {
  env: {
    browser: true,
    es2020: true,
    es6: true
  },
  globals: {
    __GLIMMER_TEMPLATE: 'readonly'
  },
  parser: '@typescript-eslint/parser',
  // extends: ['plugin:@typescript-eslint/recommended'],

  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2020,
    ecmaFeatures: {
      legacyDecorators: true
    },
    babelOptions: {
      configFile:
        '/Users/thomas/code/gossi/frontend-configs/node_modules/.pnpm/eslint-plugin-decorator-position@5.0.2_@babel+eslint-parser@7.21.3_eslint@8.38.0/node_modules/eslint-plugin-decorator-position/babel.config.cjs'
    }
  },
  plugins: ['@typescript-eslint', 'ember', 'prettier', 'decorator-position', 'simple-import-sort'],
  extends: ['plugin:@typescript-eslint/recommended'],
  rules: {
    'prettier/prettier': [
      'error',
      {
        singleQuote: true,
        trailingComma: 'none',
        printWidth: 100,
        plugins: ['prettier-plugin-ember-template-tag'],
        overrides: [
          {
            files: ['*.md', '*.hbs'],
            options: {
              singleQuote: false
            }
          }
        ]
      }
    ]
  }
};
```

Ran the script and ESLint did not complain:

```bash
pnpm eslint src/components/greeting.gts                                                                                                               ─╯
 WARN  Unsupported engine: wanted: {"node":">= v18.15.0"} (current: {"node":"v18.14.2","pnpm":"7.19.0"})
```

**Case 2**

Replaced `.eslintrc.js` with the below content using `overrides` settings:

```js
module.exports = {
  env: {
    browser: true,
    es2020: true,
    es6: true
  },
  globals: {
    __GLIMMER_TEMPLATE: 'readonly'
  },
  // root: true,
  overrides: [
    {
      parser: '@typescript-eslint/parser',
      // extends: ['plugin:@typescript-eslint/recommended'],

      parserOptions: {
        sourceType: 'module',
        ecmaVersion: 2020,
        ecmaFeatures: {
          legacyDecorators: true
        },
        babelOptions: {
          configFile:
            '/Users/thomas/code/gossi/frontend-configs/node_modules/.pnpm/eslint-plugin-decorator-position@5.0.2_@babel+eslint-parser@7.21.3_eslint@8.38.0/node_modules/eslint-plugin-decorator-position/babel.config.cjs'
        }
      },
      plugins: [
        '@typescript-eslint',
        'ember',
        'prettier',
        'decorator-position',
        'simple-import-sort'
      ],
      extends: ['plugin:@typescript-eslint/recommended'],
      rules: {
        'prettier/prettier': [
          'error',
          {
            singleQuote: true,
            trailingComma: 'none',
            printWidth: 100,
            plugins: ['prettier-plugin-ember-template-tag'],
            overrides: [
              {
                files: ['*.md', '*.hbs'],
                options: {
                  singleQuote: false
                }
              }
            ]
          }
        ]
      },
      files: ['{src,app,addon,addon-test-support,tests,types}/**/*.gts']
    }
  ]
};
```

This time ESLint did complain:

```bash
pnpm eslint src/components/greeting.gts                                                                                                               ─╯
 WARN  Unsupported engine: wanted: {"node":">= v18.15.0"} (current: {"node":"v18.14.2","pnpm":"7.19.0"})

/Users/joelmathew/Desktop/frontend-configs/testing/ember-addon-v2-gts/src/components/greeting.gts
  16:24  error  Replace `⏎····<span·...attributes>{{@hello}}·to·{{this.to}}</span>⏎··` with `<span·...attributes>{{@hello}}·to·{{this.to}}</span>`  prettier/prettier

✖ 1 problem (1 error, 0 warning
```

I tried adding the `eslint-config-pretter` shareable config to see if things work but it still complains.

Looking at the error I feel it is a problem with the `eslint-plugin-prettier` not being able to recognise the `overrides` field 
in `.eslintrc` or something like that.

```bash
error  Replace `⏎····<span·...attributes>{{@hello}}·to·{{this.to}}</span>⏎··` with `<span·...attributes>{{@hello}}·to·{{this.to}}</span>`  prettier/prettier
```

**Comment by nzakas** [comment]
I'm converting this to a discussion because 1) this error is coming from a plugin, not from ESLint itself, so you should be working with the plugin author to resolve it, and 2) we aren't making any further changes to the eslintrc config system as we switch to the new config system.

**Comment by gossi** [comment, marked as answer]
Thanks, makes sense to focus on the new config system, which I learned about today - and I am excited about.

But according to your hints and me using that to start another debugging session by finding the appropriate source I found when reading about the new config file system. I debugged into my problem and found the root cause. As eslint runs this through a parser already, it translates `gts` (with `<template>`) into plain javascript to have it understand to all the plugins, etc.

On the other side, with prettier and the [prettier-plugin-ember-template-tag](https://github.com/gitKrystan/prettier-plugin-ember-template-tag) plugin it expects the _raw_ source file. As they are different, this will always fail.

I'll talk to the original authors and proceed from there. Thanks for your time on this :)

**Reply by backspace** [reply]
Thanks for writing this up @gossi, did you ever find a resolution? I am [seeing this too](https://github.com/cardstack/boxel/actions/runs/4984782463/jobs/8923568006?pr=337#step:5:303) 😯

**Reply by gossi** [reply]
There is already an open issue for this: gitKrystan/prettier-plugin-ember-template-tag#20 - which contains more information/links about the problem.

What I did, was to deactivate `prettier/prettier` for `.gts` files until the problem above is resolved.
